src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,174 @@
+using static Microsoft.CodeAnalysis.MSBuild.BuildHostProcessManager;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+
+internal sealed class FileBasedProgramsProjectSystem : LanguageServerProjectLoader, ILspMiscellaneousFilesWorkspaceProvider
+{
+    private readonly ILspServices _lspServices;
+    private readonly ILogger<FileBasedProgramsProjectSystem> _logger;
+    private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+    public FileBasedProgramsProjectSystem(
+        ILspServices lspServices,
+        IMetadataAsSourceFileService metadataAsSourceFileService,
+        LanguageServerWorkspaceFactory workspaceFactory,
+        IFileChangeWatcher fileChangeWatcher,
+        IGlobalOptionService globalOptionService,
+        ILoggerFactory loggerFactory,
+        IAsynchronousOperationListenerProvider listenerProvider,
+        ProjectLoadTelemetryReporter projectLoadTelemetry,
+        ServerConfigurationFactory serverConfigurationFactory,
+        BinlogNamer binlogNamer)
+            : base(
+                workspaceFactory.FileBasedProgramsProjectFactory,
+                workspaceFactory.TargetFrameworkManager,
+                workspaceFactory.ProjectSystemHostInfo,
+                fileChangeWatcher,
+                globalOptionService,
+                loggerFactory,
+                listenerProvider,
+                projectLoadTelemetry,
+                serverConfigurationFactory,
+                binlogNamer)
+    {
+        _lspServices = lspServices;
+        _logger = loggerFactory.CreateLogger<FileBasedProgramsProjectSystem>();
+        _metadataAsSourceFileService = metadataAsSourceFileService;
+    }
+
+    public Workspace Workspace => ProjectFactory.Workspace;
+
+    public async Task<TextDocument?> AddMiscellaneousDocumentAsync(DocumentUri uri, SourceText documentText, string languageId, ILspLogger logger)
+    {
+        var (documentPath, isFile) = uri.ParsedUri is { } parsedUri
+            ? (ProtocolConversions.GetDocumentFilePathFromUri(parsedUri), isFile: parsedUri.IsFile)
+            : (uri.UriString, isFile: false);
+
+        var container = documentText.Container;
+        if (_metadataAsSourceFileService.TryAddDocumentToWorkspace(documentPath, container, out var documentId))
+        {
+            var metadataWorkspace = _metadataAsSourceFileService.TryGetWorkspace();
+            Contract.ThrowIfNull(metadataWorkspace);
+            var metadataDoc = metadataWorkspace.CurrentSolution.GetRequiredDocument(documentId);
+            return metadataDoc;
+        }
+
+        var languageInfoProvider = _lspServices.GetRequiredService<ILanguageInfoProvider>();
+        if (!languageInfoProvider.TryGetLanguageInformation(uri, languageId, out var languageInformation))
+        {
+            // Only log here since throwing here could take down the LSP server.
+            logger.LogError($"Could not find language information for {uri} with absolute path {documentPath}");
```

Comment:
this definitely concerns me.  i'm not a fan of this sort of approach where we're just silent, while things aren't working properly.  i'd prefer we define very strict, exacting, contracts between components, and we aggressively fail if we violate them.  otehrwise, we'll be in a state where things just aren't working, and it's unclear why.

Reply:
Note - this is existing code that moved.  So likely shouldn't be changed in this PR.  

But I don't think that necessarily applies here:
1.  We don't control all the requests that the client may make to us.  We have to be somewhat resilient to some unexpected requests.
2.  This is not a silent failure.  It is silent here because we don't want to take down the queue.  But the request will throw later when it cannot find the document.

It's totally possible for a third party extension to create an unknown (to us) document and trigger this code path.  I don't think it is reasonable to crash if that happens.

Reply:
> This is reusing the existing pattern from the other implementation. @dibarbet and I are generally agreeing we should be changing the pattern here the contract of this method being throws are OK and the queue manager will catch it. (Since we're not immune to bugs anywhere else.)

(should have refreshed before I replied).  But yes fine if this throws and is caught outside - but it should not crash.  As mentioned above, this is recoverable, and any missing documents will actually throw (and be visible to the client) later on.